spark/src/main/scala/org/apache/spark/sql/delta/ResolveDeltaIdentifier.scala

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (2023) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.delta
+
+import org.apache.hadoop.fs.Path
+import org.apache.spark.sql.{AnalysisException, SparkSession}
+import org.apache.spark.sql.catalyst.TimeTravel
+import org.apache.spark.sql.catalyst.analysis.{AnalysisErrorAt, ResolvedTable, UnresolvedTable}
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.delta.catalog.DeltaTableV2
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation
+
+case class ResolveDeltaIdentifier(sparkSession: SparkSession) extends Rule[LogicalPlan] {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperators {
+    case u: UnresolvedDeltaIdentifier =>
+      resolveDeltaIdentifier(u)
+
+    // Since the Delta's TimeTravel is a leaf node, we have to resolve the UnresolvedIdentifier.
+    case tt @ TimeTravel(unresolvedId: UnresolvedDeltaIdentifier, _, _, _) =>
+      val sourceRelation = resolveDeltaIdentifier(unresolvedId)
+      tt.copy(relation = sourceRelation)
+  }
+
+  private def resolveDeltaIdentifier(
+      unresolvedId: UnresolvedDeltaIdentifier): DataSourceV2Relation = {
+    // Resolve the identifier as a Delta table.
+    val optionalDeltaTableV2 = resolveAsTable(unresolvedId)
+
+    // If the identifier is not a Delta table, try to resolve it as a Delta file table.
+    val deltaTableV2 = optionalDeltaTableV2.getOrElse {
```

Comment:
Yes, it is mentioned in the comment of `resolveAsTable`. The short answer is that the table identifier doesn't exist in spark catalogs.
```
  // Resolve the identifier as a table in the Spark catalogs.
  // Return None if the table doesn't exist.
  // If the table exists but is not a Delta table, throw an exception.
  // If the identifier is a view, throw an exception.
```

Reply:
> If such circumstances exist, can/should we fix table resolution directly, instead of compensating here?

What do you mean by "fix table resolution directly"?
I believe there is a corner case here.
In Spark, You can create a table `/tmp/foo` under a database `delta`
```
use delta;
create table `/tmp/foo` using delta;
```
So we always have to look up as a catalog table, and try loading it as a path if the table doesn't exist.

Reply:
<img width="655" alt="image" src="https://github.com/delta-io/delta/assets/1097932/f98cc7e3-9bb2-4a93-a3ac-4aa36a2e5a55">